ouroboros-consensus-test/src/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,638 @@
+    -- sd = (Mat.! 0) $ stimes (2^(10 :: Int) :: Int) p
+-}
+
+    -- Transition matrix of an Markov Chain where each state is the
+    -- (idealized?) number of active slots in an EHCG window.
+    --
+    -- (This is a /stochastic matrix/ because every /row/ sums to 1. And every
+    -- state is reachable from every other, since @0 < asc@.)
+    --
+    -- We have one lingering doubt about whether this system perfectly models
+    -- the behavior of the loop in 'uniformTheHonestChain' that visits the
+    -- remaining windows, hence the "idealized?" above. In particular, these
+    -- transitions are the result of assuming that the active slots within a
+    -- window are uniformly distributed after it has been visited by that loop.
+    -- Specifically, we assume that the probability of the oldest slot being
+    -- active is the population count divided by the window width. Our
+    -- lingering doubt is that we're overestimating the degrees of freedom such
+    -- that somehow a slot's value from /before/ some window is unexpectedly
+    -- influencing the value of the first slot in that window in a way that is
+    -- not accounted for by the number of active slots in the preceding window,
+    -- ie that the Markov Property does not actually hold. I think the seed for
+    -- this doubt was Esgen's earlier insight that a chain with /exactly/ @n@
+    -- actives in every @d@ slot window is periodic. In that case, if there are
+    -- more than @d@ contiguous windows that each have exactly @n@ actives,
+    -- then there is perfect auto-correlation throughout those windows, and it
+    -- spans farther than one window. That's our suspicion for the kind of
+    -- counterexample of the Markov Property of this transition matrix.
+    --
+    -- That said, even if it's not a perfect model, we would guess it's usually
+    -- a fine approximation to use for sampling the density of the first
+    -- window.
+    --
+    -- The indices @i@ and @j@ are @n@ less than state they represent. In other
+    -- words, they count the /extra/ active slots, not just the active slots.
+    p = Mat.build (sz, sz) $ curry $ \case
+      (0, 0) -> (dd - dd * f + nn * f) / dd
+      (0, 1) -> (     dd * f - nn * f) / dd
+
+      (i, j)
+        | dd - nn == i, j == i - 1
+        -> 1 - f
+        | dd - nn == i, j == i
+        -> f
+
+        | j == i - 1 -> (x - x * f)                   / dd
+        | j == i     -> (2 * x * f + dd - dd * f - x) / dd
+        | j == i + 1 -> (dd * f - x * f)              / dd
+        where
+          x = i + nn
+
+      _ -> 0
+
+{-
+
+Derivation of the transition matrix used in 'stationaryWindowSize'.
+The x/d summand is the chance we're sliding the window off of an active slot.
+The (d - x)/d summand is the chance we're sliding the window off of an empty slot.
+
+One reflecting barrier at n.
+
+    P( n -> n     ) = (n/d) * 1         +   ((d - n)/d) * 1
```

Comment:
Something is wrong here, probably replace the last $1$ by $1-f$.

Reply:
Fixed 👍